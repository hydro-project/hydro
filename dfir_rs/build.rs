use rustc_version::{Channel, version_meta};

fn main() {
    println!("cargo:rerun-if-changed=build.rs");
    println!("cargo::rustc-check-cfg=cfg(nightly)");
    if matches!(
        version_meta().map(|meta| meta.channel),
        Ok(Channel::Nightly)
    ) {
        println!("cargo:rustc-cfg=nightly");
    }
}
