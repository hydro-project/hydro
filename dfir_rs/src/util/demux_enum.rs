//! Trait for the `demux_enum` derive and operator.

use std::task::{Context, Poll};

pub use dfir_macro::DemuxEnum;

/// Trait for use with the `demux_enum` operator.
///
/// This trait is meant to be derived: `#[derive(DemuxEnum)]`.
///
/// The derive will implement this such that `Outputs` can be any tuple where each item is a
/// `Sink` that corresponds to each of the variants of the tuple, in alphabetic order.
#[diagnostic::on_unimplemented(
    note = "ensure there is exactly one output for each enum variant.",
    note = "ensure that the type for each output is a tuple of the field for the variant: `()`, `(a,)`, or `(a, b, ...)`."
)]
pub trait DemuxEnumSink<Outputs>: DemuxEnumBase {
    /// The error type for pushing self into the `Outputs` `Sink`s.
    type Error;

    /// Call `poll_ready` on all `Outputs`.
    fn poll_ready(outputs: &mut Outputs, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>>;

    /// Pushes `self` into the corresponding output sink in `outputs`.
    fn start_send(self, outputs: &mut Outputs) -> Result<(), Self::Error>;

    /// Call `poll_flush` on all `Outputs`.
    fn poll_flush(outputs: &mut Outputs, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>>;

    /// Call `poll_close` on all `Outputs`.
    fn poll_close(outputs: &mut Outputs, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>>;
}

/// Special case of [`DemuxEnum`] for when there is only one variant.
#[diagnostic::on_unimplemented(
    note = "requires that the enum have only one variant.",
    note = "ensure there are no missing outputs; there must be exactly one output for each enum variant."
)]
pub trait SingleVariant: DemuxEnumBase {
    /// Output tuple type.
    type Output;
    /// Convert self into it's single variant tuple Output.
    fn single_variant(self) -> Self::Output;
}

/// Base implementation to constrain that [`DemuxEnum<SOMETHING>`] is implemented.
#[diagnostic::on_unimplemented(note = "use `#[derive(dfir_rs::DemuxEnum)]`")]
pub trait DemuxEnumBase {}
