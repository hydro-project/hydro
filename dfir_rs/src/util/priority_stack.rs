//! A priority queue in which elements of the same priority are popped in a LIFO order.

use smallvec::SmallVec;

/// A priority stack in which elements of the same priority are popped in a LIFO order.
// TODO(mingwei): Keep an upper bound on current priority to avoid scanning all stacks?
#[derive(Debug, Clone)]
pub struct PriorityStack<T> {
    /// Note: inner stack `Vec`s may be empty.
    stacks: Vec<SmallVec<[T; 1]>>,
}

impl<T> PriorityStack<T> {
    /// Creates a new, empty `PriorityStack`.
    pub fn new() -> Self {
        Self {
            stacks: Vec::default(),
        }
    }

    /// Creates a new, empty `PriorityStack` with pre-allocated capacity up to the given priority.
    pub fn with_priority_capacity(priority: usize) -> Self {
        Self {
            stacks: Vec::with_capacity(priority),
        }
    }

    /// Pushes an element onto the stack with the given priority.
    pub fn push(&mut self, priority: usize, item: T) {
        if priority >= self.stacks.len() {
            self.stacks.resize_with(priority + 1, Default::default);
        }
        self.stacks[priority].push(item);
    }

    /// Pops an element from the stack with the highest priority.
    pub fn pop(&mut self) -> Option<T> {
        self.stacks
            .iter_mut()
            .rev()
            .filter_map(SmallVec::pop)
            .next()
    }

    /// Pops an element from the stack and return `(priority, item)`.
    pub fn pop_prio(&mut self) -> Option<(usize, T)> {
        self.stacks
            .iter_mut()
            .enumerate()
            .rev()
            .filter_map(|(i, stack)| stack.pop().map(|x| (i, x)))
            .next()
    }

    /// Returns the item with the highest priority without removing it.
    pub fn peek(&self) -> Option<&T> {
        self.stacks
            .iter()
            .rev()
            .filter_map(|stack| stack.last())
            .next()
    }

    /// Returns the item with the highest priority and its priority without removing it.
    pub fn peek_prio(&self) -> Option<(usize, &T)> {
        self.stacks
            .iter()
            .enumerate()
            .rev()
            .filter_map(|(i, stack)| stack.last().map(|x| (i, x)))
            .next()
    }

    /// Returns the number of elements in the `PriorityStack`.
    pub fn len(&self) -> usize {
        self.stacks.iter().map(SmallVec::len).sum()
    }

    /// Returns true if the `PriorityStack` is empty.
    pub fn is_empty(&self) -> bool {
        self.stacks.is_empty()
    }
}

impl<T> Default for PriorityStack<T> {
    fn default() -> Self {
        Self::new()
    }
}

impl<T> Extend<(usize, T)> for PriorityStack<T> {
    fn extend<I: IntoIterator<Item = (usize, T)>>(&mut self, iter: I) {
        for (priority, item) in iter {
            self.push(priority, item);
        }
    }
}
