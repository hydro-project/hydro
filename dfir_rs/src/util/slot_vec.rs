//! A Vec-based SlotMap-esque datastructure and corresponding Key type.

use std::fmt::{Debug, Display, Formatter};
use std::hash::{Hash, Hasher};
use std::marker::PhantomData;
use std::ops::{Index, IndexMut};

/// A key into a SlotVec.
#[repr(transparent)]
pub struct Key<Tag: ?Sized> {
    index: usize,
    _phantom: PhantomData<Tag>,
}
impl<Tag: ?Sized> Key<Tag> {
    /// Creates a Key from a raw index. Avoid using this function directly.
    pub fn from_raw(index: usize) -> Self {
        Key {
            index,
            _phantom: PhantomData,
        }
    }
}
impl<Tag: ?Sized> Clone for Key<Tag> {
    fn clone(&self) -> Self {
        *self
    }
}
impl<Tag: ?Sized> Copy for Key<Tag> {}
impl<Tag: ?Sized> PartialOrd for Key<Tag> {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}
impl<Tag: ?Sized> Ord for Key<Tag> {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.index.cmp(&other.index)
    }
}
impl<Tag: ?Sized> PartialEq for Key<Tag> {
    fn eq(&self, other: &Self) -> bool {
        self.index == other.index
    }
}
impl<Tag: ?Sized> Eq for Key<Tag> {}
impl<Tag: ?Sized> Hash for Key<Tag> {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.index.hash(state);
    }
}
impl<Tag: ?Sized> Debug for Key<Tag> {
    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
        write!(f, "Key({})", self.index)
    }
}
impl<Tag: ?Sized> Display for Key<Tag> {
    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
        write!(f, "{}", self.index)
    }
}

/// A Vec-based SlotMap-esque datastructure without removes.
///
/// Analogous to [`slotmap::SlotMap`], but avoids the overhead of tracking removed keys.
#[repr(transparent)]
pub struct SlotVec<Tag: ?Sized, Val> {
    slots: Vec<Val>,
    _phantom: PhantomData<Tag>,
}
impl<Tag: ?Sized, Val> SlotVec<Tag, Val> {
    /// Creates a new `SlotVec`.
    pub fn new() -> Self {
        Self {
            slots: Vec::default(),
            _phantom: PhantomData,
        }
    }

    /// Inserts a value into the `SlotVec` and returns the key.
    pub fn insert(&mut self, value: Val) -> Key<Tag> {
        let key = Key::from_raw(self.slots.len());
        self.slots.push(value);
        key
    }

    /// Use the provided function to generate a value given the key and insert it into the `SlotVec`.
    pub fn insert_with_key<F>(&mut self, func: F) -> Key<Tag>
    where
        F: FnOnce(Key<Tag>) -> Val,
    {
        let key = Key::from_raw(self.slots.len());
        self.slots.push((func)(key));
        key
    }

    /// Returns a reference to the value associated with the key.
    pub fn get(&self, key: Key<Tag>) -> Option<&Val> {
        self.slots.get(key.index)
    }

    /// Returns a mutable reference to the value associated with the key.
    pub fn get_mut(&mut self, key: Key<Tag>) -> Option<&mut Val> {
        self.slots.get_mut(key.index)
    }

    /// Returns the number of elements in the `SlotVec`.
    pub fn len(&self) -> usize {
        self.slots.len()
    }

    /// Returns true if the `SlotVec` is empty.
    pub fn is_empty(&self) -> bool {
        self.slots.is_empty()
    }
}
impl<Tag: ?Sized, Val> Index<Key<Tag>> for SlotVec<Tag, Val> {
    type Output = Val;

    fn index(&self, key: Key<Tag>) -> &Self::Output {
        self.get(key).unwrap()
    }
}
impl<Tag: ?Sized, Val> IndexMut<Key<Tag>> for SlotVec<Tag, Val> {
    fn index_mut(&mut self, key: Key<Tag>) -> &mut Self::Output {
        self.get_mut(key).unwrap()
    }
}
impl<Key: ?Sized, Val> Default for SlotVec<Key, Val> {
    fn default() -> Self {
        Self::new()
    }
}

/// A secondary map used to associated data with keys from elements in an existing [`SlotVec`].
///
/// Analogous to [`slotmap::SecondaryMap`].
pub struct SecondarySlotVec<Tag: ?Sized, Val> {
    slots: Vec<Option<Val>>,
    _phantom: PhantomData<Tag>,
}
impl<Tag: ?Sized, Val> SecondarySlotVec<Tag, Val> {
    /// Creates a new `SecondarySlotVec`.
    pub fn new() -> Self {
        Self {
            slots: Vec::default(),
            _phantom: PhantomData,
        }
    }

    /// Inserts a value into the `SecondarySlotVec` and returns the previous value associated with the key.
    pub fn insert(&mut self, key: Key<Tag>, value: Val) -> Option<Val> {
        if key.index >= self.slots.len() {
            self.slots.resize_with(key.index + 1, || None);
        }
        self.slots[key.index].replace(value)
    }

    /// Removes a value associated with the key from the `SecondarySlotVec` and returns it.
    pub fn remove(&mut self, key: Key<Tag>) -> Option<Val> {
        // TODO(mingwei): Shrink the vector?
        self.slots[key.index].take()
    }

    /// Returns a reference to the value associated with the key.
    pub fn get(&self, key: Key<Tag>) -> Option<&Val> {
        self.slots.get(key.index).and_then(|v| v.as_ref())
    }

    /// Returns a mutable reference to the value associated with the key.
    pub fn get_mut(&mut self, key: Key<Tag>) -> Option<&mut Val> {
        self.slots.get_mut(key.index).and_then(|v| v.as_mut())
    }
}
impl<Tag: ?Sized, Val> Default for SecondarySlotVec<Tag, Val> {
    fn default() -> Self {
        Self::new()
    }
}
impl<Tag: ?Sized, Val> Index<Key<Tag>> for SecondarySlotVec<Tag, Val> {
    type Output = Val;

    fn index(&self, key: Key<Tag>) -> &Self::Output {
        self.get(key).unwrap()
    }
}
impl<Tag: ?Sized, Val> IndexMut<Key<Tag>> for SecondarySlotVec<Tag, Val> {
    fn index_mut(&mut self, key: Key<Tag>) -> &mut Self::Output {
        self.get_mut(key).unwrap()
    }
}
