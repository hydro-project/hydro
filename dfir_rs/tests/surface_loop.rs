use dfir_rs::util::{collect_ready, iter_batches_stream};
use dfir_rs::{assert_graphvis_snapshots, dfir_syntax};
use multiplatform_test::multiplatform_test;

#[multiplatform_test(test, env_tracing, wasm)]
pub fn test_flo_syntax() {
    let (result_send, mut result_recv) = dfir_rs::util::unbounded_channel::<_>();

    let mut df = dfir_syntax! {
        users = source_iter(["alice", "bob"]);
        messages = source_stream(iter_batches_stream(0..12, 3));
        loop {
            users -> prefix() -> [0]cp;
            messages -> batch() -> [1]cp;
            cp = cross_join()
                -> map(|item| (context.loop_iter_count(), item))
                -> for_each(|x| result_send.send(x).unwrap());
        };
    };
    assert_graphvis_snapshots!(df);
    df.run_available();

    assert_eq!(
        &[
            (0, ("alice", 0)),
            (0, ("alice", 1)),
            (0, ("alice", 2)),
            (0, ("bob", 0)),
            (0, ("bob", 1)),
            (0, ("bob", 2)),
            (1, ("alice", 3)),
            (1, ("alice", 4)),
            (1, ("alice", 5)),
            (1, ("bob", 3)),
            (1, ("bob", 4)),
            (1, ("bob", 5)),
            (2, ("alice", 6)),
            (2, ("alice", 7)),
            (2, ("alice", 8)),
            (2, ("bob", 6)),
            (2, ("bob", 7)),
            (2, ("bob", 8)),
            (3, ("alice", 9)),
            (3, ("alice", 10)),
            (3, ("alice", 11)),
            (3, ("bob", 9)),
            (3, ("bob", 10)),
            (3, ("bob", 11)),
        ],
        &*collect_ready::<Vec<_>, _>(&mut result_recv)
    );
}

#[multiplatform_test(test, wasm, env_tracing)]
pub fn test_flo_nested() {
    let (result_send, mut result_recv) = dfir_rs::util::unbounded_channel::<_>();

    let mut df = dfir_syntax! {
        users = source_iter(["alice", "bob"]);
        messages = source_stream(iter_batches_stream(0..12, 3));
        loop {
            users -> prefix() -> [0]cp;
            messages -> batch() -> [1]cp;
            cp = cross_join();
            loop {
                cp
                    -> all_once()
                    -> map(|item| (context.current_tick().0, item))
                    -> for_each(|x| result_send.send(x).unwrap());
            };
        };
    };
    assert_graphvis_snapshots!(df);
    df.run_available();

    assert_eq!(
        &[
            (0, ("alice", 0)),
            (0, ("alice", 1)),
            (0, ("alice", 2)),
            (0, ("bob", 0)),
            (0, ("bob", 1)),
            (0, ("bob", 2)),
            (1, ("alice", 3)),
            (1, ("alice", 4)),
            (1, ("alice", 5)),
            (1, ("bob", 3)),
            (1, ("bob", 4)),
            (1, ("bob", 5)),
            (2, ("alice", 6)),
            (2, ("alice", 7)),
            (2, ("alice", 8)),
            (2, ("bob", 6)),
            (2, ("bob", 7)),
            (2, ("bob", 8)),
            (3, ("alice", 9)),
            (3, ("alice", 10)),
            (3, ("alice", 11)),
            (3, ("bob", 9)),
            (3, ("bob", 10)),
            (3, ("bob", 11)),
        ],
        &*collect_ready::<Vec<_>, _>(&mut result_recv)
    );
}

#[multiplatform_test(test, env_tracing)]
pub fn test_flo_repeat_n() {
    let (result_send, mut result_recv) = dfir_rs::util::unbounded_channel::<_>();

    let mut df = dfir_syntax! {
        users = source_iter(["alice", "bob"]);
        messages = source_stream(iter_batches_stream(0..9, 3));
        loop {
            users -> prefix() -> [0]cp;
            messages -> batch() -> [1]cp;
            cp = cross_join();
            loop {
                cp -> repeat_n(2)
                    -> inspect(|x| println!("{:?} {}", x, context.loop_iter_count()))
                    -> for_each(|x| result_send.send(x).unwrap());
            };
        };
    };
    assert_graphvis_snapshots!(df);
    df.run_available();

    assert_eq!(
        &[
            ("alice", 0),
            ("alice", 1),
            ("alice", 2),
            ("bob", 0),
            ("bob", 1),
            ("bob", 2),
            ("alice", 0),
            ("alice", 1),
            ("alice", 2),
            ("bob", 0),
            ("bob", 1),
            ("bob", 2),
            ("alice", 3),
            ("alice", 4),
            ("alice", 5),
            ("bob", 3),
            ("bob", 4),
            ("bob", 5),
            ("alice", 3),
            ("alice", 4),
            ("alice", 5),
            ("bob", 3),
            ("bob", 4),
            ("bob", 5),
            ("alice", 6),
            ("alice", 7),
            ("alice", 8),
            ("bob", 6),
            ("bob", 7),
            ("bob", 8),
            ("alice", 6),
            ("alice", 7),
            ("alice", 8),
            ("bob", 6),
            ("bob", 7),
            ("bob", 8),
        ],
        &*collect_ready::<Vec<_>, _>(&mut result_recv)
    );
}

#[multiplatform_test(test, wasm, env_tracing)]
pub fn test_flo_repeat_n_nested() {
    let (result_send, mut result_recv) = dfir_rs::util::unbounded_channel::<_>();

    let mut df = dfir_syntax! {
        usrs1 = source_iter(["alice", "bob"]);
        loop {
            usrs2 = usrs1 -> batch();
            loop {
                usrs3 = usrs2 -> repeat_n(3) -> inspect(|x| println!("A {:?} {}", x, context.loop_iter_count()));
                loop {
                    usrs3 -> repeat_n(3)
                        -> inspect(|x| println!("B {:?} {}", x, context.loop_iter_count()))
                        -> for_each(|x| result_send.send(x).unwrap());
                };
            };
        };
    };
    assert_graphvis_snapshots!(df);
    df.run_available();

    assert_eq!(
        &[
            "alice", "bob", "alice", "bob", "alice", "bob", "alice", "bob", "alice", "bob",
            "alice", "bob", "alice", "bob", "alice", "bob", "alice", "bob",
        ],
        &*collect_ready::<Vec<_>, _>(&mut result_recv)
    );
}

#[multiplatform_test]
pub fn test_flo_repeat_n_multiple_nested() {
    let (result1_send, mut result1_recv) = dfir_rs::util::unbounded_channel::<_>();
    let (result2_send, mut result2_recv) = dfir_rs::util::unbounded_channel::<_>();

    let mut df = dfir_syntax! {
        usrs1 = source_iter(["alice", "bob"]);
        loop {
            usrs2 = usrs1 -> batch();
            loop {
                usrs3 = usrs2 -> repeat_n(3)
                    -> inspect(|x| println!("{:?} {}", x, context.loop_iter_count()))
                    -> tee();
                loop {
                    usrs3 -> repeat_n(3)
                    -> inspect(|x| println!("{} {:?} {}", line!(), x, context.loop_iter_count()))
                    -> for_each(|x| result1_send.send(x).unwrap());
                };
                loop {
                    usrs3 -> repeat_n(3)
                        -> inspect(|x| println!("{} {:?} {}", line!(), x, context.loop_iter_count()))
                        -> for_each(|x| result2_send.send(x).unwrap());
                };
            };
        };
    };
    assert_graphvis_snapshots!(df);
    df.run_available();

    assert_eq!(
        &[
            "alice", "bob", "alice", "bob", "alice", "bob", "alice", "bob", "alice", "bob",
            "alice", "bob", "alice", "bob", "alice", "bob", "alice", "bob",
        ],
        &*collect_ready::<Vec<_>, _>(&mut result1_recv)
    );

    assert_eq!(
        &[
            "alice", "bob", "alice", "bob", "alice", "bob", "alice", "bob", "alice", "bob",
            "alice", "bob", "alice", "bob", "alice", "bob", "alice", "bob",
        ],
        &*collect_ready::<Vec<_>, _>(&mut result2_recv)
    );
}

#[multiplatform_test(test, wasm, env_tracing)]
pub fn test_flo_repeat_kmeans() {
    const POINTS: &[[i32; 2]] = &[
        [-210, -104],
        [-226, -143],
        [-258, -119],
        [-331, -129],
        [-250, -69],
        [-202, -113],
        [-222, -133],
        [-232, -155],
        [-220, -107],
        [-159, -109],
        [-49, 57],
        [-156, 52],
        [-22, 125],
        [-140, 168],
        [-118, 89],
        [-93, 133],
        [-101, 80],
        [-145, 79],
        [187, 36],
        [208, -66],
        [142, 5],
        [232, 41],
        [91, -37],
        [132, 16],
        [248, -39],
        [158, 65],
        [108, -41],
        [171, -121],
        [147, 5],
        [192, 58],
    ];
    const CENTROIDS: &[[i32; 2]] = &[[-50, 0], [0, 0], [50, 0]];

    let (result_send, mut result_recv) = dfir_rs::util::unbounded_channel::<_>();

    let mut df = dfir_syntax! {
        init_points = source_iter(POINTS) -> map(std::clone::Clone::clone);
        init_centroids = source_iter(CENTROIDS) -> map(std::clone::Clone::clone);
        loop {
            batch_points = init_points -> batch();
            batch_centroids = init_centroids -> batch();

            loop {
                points = batch_points
                    -> repeat_n(10)
                    -> [0]cj;
                batch_centroids -> all_once() -> centroids;

                centroids = union() -> tee();
                centroids -> [1]cj;

                cj = cross_join_multiset()
                    -> map(|(point, centroid): ([i32; 2], [i32; 2])| {
                        let dist2 = (point[0] - centroid[0]).pow(2) + (point[1] - centroid[1]).pow(2);
                        (point, (dist2, centroid))
                    })
                    -> reduce_keyed(|(a_dist2, a_centroid), (b_dist2, b_centroid)| {
                        if b_dist2 < *a_dist2 {
                            *a_dist2 = b_dist2;
                            *a_centroid = b_centroid;
                        }
                    })
                    -> map(|(point, (_dist2, centroid))| {
                        (centroid, (point, 1))
                    })
                    -> reduce_keyed(|(p1, n1), (p2, n2): ([i32; 2], i32)| {
                        p1[0] += p2[0];
                        p1[1] += p2[1];
                        *n1 += n2;
                    })
                    -> map(|(_centroid, (p, n)): (_, ([i32; 2], i32))| {
                         [p[0] / n, p[1] / n]
                    })
                    -> next_iteration()
                    -> inspect(|x| println!("centroid: {:?}", x))
                    -> centroids;
            };

            centroids
                -> all_iterations()
                -> for_each(|x| result_send.send(x).unwrap());
        };
    };
    assert_graphvis_snapshots!(df);
    df.run_available();

    let mut result = collect_ready::<Vec<_>, _>(&mut result_recv);
    let n = result.len();
    let last = &mut result[n - 3..];
    last.sort_unstable();
    assert_eq!(&[[-231, -118], [-103, 97], [168, -6]], last);
}

#[test]
fn test_state_codegen() {
    let mut df = dfir_syntax! {
        a = source_iter((0..10).chain(5..15)) -> tee();
        pairs = a -> map(|n| (n / 3, n)) -> tee();
        loop {
            a -> batch() -> fold(|| 0, |old: &mut _, val| {
                *old += val;
            }) -> for_each(|v| println!("fold1 {:?}", v));

            a -> batch() -> fold::<'none>(|| 0, |old: &mut _, val| {
                *old += val;
            }) -> for_each(|v| println!("fold2 {:?}", v));

            a -> batch() -> reduce::<'none>(|old: &mut _, val| {
                *old += val;
            }) -> for_each(|v| println!("reduce {:?}", v));

            pairs -> batch() -> fold_keyed(|| 0, |old: &mut _, val| {
                *old += val;
            }) -> for_each(|v| println!("fold_keyed {:?}", v));

            a -> batch() -> unique() -> for_each(|v| println!("unique {:?}", v));

            j = join() -> for_each(|v| println!("join {:?}", v));
            pairs -> batch() -> [0]j;
            pairs -> batch() -> [1]j;

            aj = difference() -> for_each(|v| println!("difference {:?}", v));
            a -> batch() -> filter(|n| 0 == n % 2) -> [neg]aj;
            a -> batch() -> [pos]aj;
        };
    };
    df.run_available();
}
