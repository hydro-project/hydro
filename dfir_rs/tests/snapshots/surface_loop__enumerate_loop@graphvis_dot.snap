---
source: dfir_rs/tests/surface_loop.rs
expression: df.meta_graph().unwrap().to_dot(cfg)
---
digraph {
    node [fontname="Monaco,Menlo,Consolas,&quot;Droid Sans Mono&quot;,Inconsolata,&quot;Courier New&quot;,monospace", style=filled];
    edge [fontname="Monaco,Menlo,Consolas,&quot;Droid Sans Mono&quot;,Inconsolata,&quot;Courier New&quot;,monospace"];
    n1v1 [label="(n1v1) source_iter(0..5)", shape=invhouse, fillcolor="#88aaff"]
    n2v1 [label="(n2v1) batch()", shape=invhouse, fillcolor="#88aaff"]
    n3v1 [label="(n3v1) tee()", shape=house, fillcolor="#ffff88"]
    n4v1 [label="(n4v1) repeat_n(3)", shape=invhouse, fillcolor="#88aaff"]
    n5v1 [label="(n5v1) enumerate::<'none>()", shape=invhouse, fillcolor="#88aaff"]
    n6v1 [label="(n6v1) for_each(|x| result1_send.send(x).unwrap())", shape=house, fillcolor="#ffff88"]
    n7v1 [label="(n7v1) repeat_n(3)", shape=invhouse, fillcolor="#88aaff"]
    n8v1 [label="(n8v1) enumerate::<'loop>()", shape=invhouse, fillcolor="#88aaff"]
    n9v1 [label="(n9v1) for_each(|x| result2_send.send(x).unwrap())", shape=house, fillcolor="#ffff88"]
    n10v1 [label="(n10v1) handoff", shape=parallelogram, fillcolor="#ddddff"]
    n11v1 [label="(n11v1) handoff", shape=parallelogram, fillcolor="#ddddff"]
    n12v1 [label="(n12v1) handoff", shape=parallelogram, fillcolor="#ddddff"]
    n2v1 -> n3v1
    n1v1 -> n10v1
    n5v1 -> n6v1
    n4v1 -> n5v1
    n3v1 -> n11v1
    n8v1 -> n9v1
    n7v1 -> n8v1
    n3v1 -> n12v1
    n10v1 -> n2v1
    n11v1 -> n4v1
    n12v1 -> n7v1
    subgraph sg_1v1 {
        cluster=true
        fillcolor="#dddddd"
        style=filled
        label = "sg_1v1\nstratum 0"
        subgraph sg_1v1_var_init {
            cluster=true
            label="var init"
            n1v1
        }
    }
    subgraph loop_1v1 {
        cluster=true
        fillcolor="#ee88ee"
        style=filled
        label = loop_1v1
        subgraph sg_2v1 {
            cluster=true
            fillcolor="#dddddd"
            style=filled
            label = "sg_2v1\nstratum 1"
            subgraph sg_2v1_var_batch_init {
                cluster=true
                label="var batch_init"
                n2v1
                n3v1
            }
        }
    }
    subgraph loop_2v1 {
        cluster=true
        fillcolor="#ee88ee"
        style=filled
        label = loop_2v1
        subgraph sg_3v1 {
            cluster=true
            fillcolor="#dddddd"
            style=filled
            label = "sg_3v1\nstratum 2"
            n4v1
            n5v1
            n6v1
        }
        subgraph sg_4v1 {
            cluster=true
            fillcolor="#dddddd"
            style=filled
            label = "sg_4v1\nstratum 2"
            n7v1
            n8v1
            n9v1
        }
    }
}
